### ISSUE TITLE
gimme does not produce a goroot if /usr/local/go is removed

### ISSUE THREAD

**Issue by fkautz** [travis-ci/gimme#14]
Steps to reproduce:
1. install go to /usr/local/go
2. install and use gimme (e.g. gimme 1.4.2)
3. remove /usr/local/go
4. use gimme 1.4.2 (no goroot is produced)

Workaround:
Remove /usr/local/go and $HOME/.gimme directory. gimme will start emitting GOROOT properly.

**Comment by tianon** [comment]
Hmm, I can't seem to reproduce this:

``` console
$ docker run -it --rm golang:1.5
root@4998150c445a:/go# go env GOROOT
/usr/local/go
root@4998150c445a:/go# curl -fsSL 'https://raw.githubusercontent.com/travis-ci/gimme/master/gimme' -o /usr/local/bin/gimme
root@4998150c445a:/go# chmod +x /usr/local/bin/gimme
root@4998150c445a:/go# eval "$(gimme 1.4.3)"
go version go1.4.3 linux/amd64
root@4998150c445a:/go# go env GOROOT
/root/.gimme/versions/go1.4.3.linux.amd64
root@4998150c445a:/go# rm -rf /usr/local/go
root@4998150c445a:/go# go env GOROOT
/root/.gimme/versions/go1.4.3.linux.amd64
root@4998150c445a:/go# eval "$(gimme 1.4.3)"
go version go1.4.3 linux/amd64
root@4998150c445a:/go# go env GOROOT
/root/.gimme/versions/go1.4.3.linux.amd64
```

**Comment by meatballhat** [comment]
I can't reproduce it either 🙀  @fkautz Is this still an issue for you?

**Comment by fkautz** [comment]
I can still reproduce it. I'll create a script you can run in a docker container to reproduce the behavior.

``` sh
$ go version
go: cannot find GOROOT directory: /usr/local/go
```

**Comment by fkautz** [comment]
Compare the broken state:

``` sh
# gimme 1.6.2

unset GOOS
unset GOARCH
unset GOROOT
export PATH="/root/.gimme/versions/go1.6.2.linux.amd64/bin:${PATH}"
go version >&2
```

vs removing ~/.gimme and rerunning:

``` sh
# gimme 1.6.2

unset GOOS
unset GOARCH
export GOROOT="/root/.gimme/versions/go1.6.2.linux.amd64"
export PATH="/root/.gimme/versions/go1.6.2.linux.amd64/bin:${PATH}"
go version >&2

export GIMME_ENV="/root/.gimme/envs/go1.6.2.env"
```

**Comment by fkautz** [comment]
https://github.com/fkautz/gimme-broken-example

**Comment by meatballhat** [comment]
@fkautz Thanks for the example!  I think the case you're hitting was introduced when we started checking for existing installations for both `GIMME_TYPE=auto` and `GIMME_TYPE=binary`.  My current opinion is that if the state of the system `GOROOT` changes after the initial gimme env file is generated, then a broken installation is correct behavior.  What do you think about addressing this in the docs such that if a telltale `go: cannot find GOROOT directory: /usr/local/go` occurs, the recommendation is to `rm -rf ~/.gimme/envs` and retry?

**Comment by fkautz** [comment]
I'm fine with that, long term would probably be good to fix in script, e..g detect and do the right thing or detect and recommend removing ~/.gimme. I don't think most CI users will hit this, but for power users who wish to switch between go versions quickly, knowing this caveat would definitely help.

**Comment by dotdoom** [comment]
Another problem on CI is that when go 1.6 is installed (which expects [context](https://golang.org/pkg/context/) in `import "golang.org/x/net/context"`) and go 1.7 is installed (expecting `import "context"`), then `$GOROOT` matters and must point to 1.7, otherwise sources can not compile complaining that `$GOROOT/src/context` is not there.

Same may apply to `$GOTOOLDIR` at some point, when a forward-incompatible change is introduced.

**Comment by philpennock** [comment]
I've pushed a fix such that `gimme` always sets its own `GOROOT` in the eval'd variables, which _should_ avoid the problem here, if I'm understanding it right.

Assuming that 913dd17f is not reverted then, with an updated `gimme`, is this problem still reproducible?

**Comment by philpennock** [comment]
It's been 10 months and nobody has been able to reproduce since 913dd17 was merged.  The changed functionality was included in release 1.2.0.

I'm marking this as resolved.  Please re-open if it can still be reproduced with current `gimme`.
